File: iOS/Code/Swift/bugTrap/bugTrapKit/Controllers/ViewControllers/Views/BtAnnotateImageView.swift
//
//  BtAnnotateImageView.swift
//  bugTrap
//
//  Created by Colby L Williams on 6/30/14.
//  Copyright (c) 2014 bugTrap. All rights reserved.
//

import UIKit
import CoreText

class BtAnnotateImageView: UIView {
	
	var tool = Annotate.Tool.Marker
	

	let annotateUndoManager = NSUndoManager()
	
	var annotationDataCacheHasChangesSinceCHecked = false
	
	var annotationDataCache: [AnnotationData] = [AnnotationData]() {
		willSet {
			annotationDataCacheHasChangesSinceCHecked = true
		}
	}
	
	var annotationDataCacheArchive = [AnnotationData]()
	
	
	var strokeColor = Annotate.Color.Black.color()
	
	var bzPath = UIBezierPath()
	
	var arrowPoints = [CGPoint](count: 7, repeatedValue: CGPointZero)
	
	var markerPoints = [CGPoint](count: 5, repeatedValue: CGPointZero)
	
	var pathStart = CGPointZero
	
	var counter = 0, calloutCount = 0
	
	var incrementImage: UIImage?, bufferImage: UIImage?
	
	var bufferImageView: UIImageView?
	
	var touchesDidMove = false
	
	var displayTransform : CGAffineTransform?
	
	var displayRect = CGRectZero, contextRect = CGRectZero, bufferRect = CGRectZero
	
	var isRedrawingAfterOrientationChange = false
	
	var imageHasChanges: Bool {
		let changed = annotationDataCache.isEmpty ? false : annotationDataCacheHasChangesSinceCHecked
		annotationDataCacheHasChangesSinceCHecked = false
		return changed
	}
	
	func resetContextAndState() {
		
		annotateUndoManager.removeAllActionsWithTarget(self)
		annotationDataCache.removeAll(keepCapacity: false)
		annotationDataCacheArchive.removeAll(keepCapacity: false)
		bzPath.removeAllPoints()
		arrowPoints = [CGPoint](count: 7, repeatedValue: CGPointZero)
		markerPoints = [CGPoint](count: 5, repeatedValue: CGPointZero)
		pathStart = CGPointZero
		counter = 0
		incrementImage = nil
		touchesDidMove = false
		displayTransform = nil
		displayRect = CGRectZero
		contextRect = CGRectZero
		bufferRect  = CGRectZero
	}
	

	func updateContextAndStateForRotation() {
		displayRect = CGRectZero
		contextRect = CGRectZero
		bufferRect  = CGRectZero
		bufferImageView?.removeFromSuperview()
	}
	

	func refreshLayoutAndImageForRotation() {
		if !TrapState.Shared.hasActiveSnapshotImage {
			return
		}
		
		isRedrawingAfterOrientationChange = true
		
		redrawBitmap()
	}
	
	required init(coder aDecoder: NSCoder) {
		
		super.init(coder: aDecoder)
		
		multipleTouchEnabled = false
		backgroundColor = UIColor.whiteColor()
		
		configureStroke(Annotate.Color.Black.color(), width: 4.0)
	}
	
	
	
	func configureTool (annotateTool: Annotate.Tool) {
		tool = annotateTool
	}
	
	
	
	func configureStroke (color: UIColor, width: CGFloat = 0.0) {
		
		strokeColor = color
		
		if width > 0.0 {
			bzPath.lineWidth = width
		}
	}
	
	
	
	func refreshLayoutAndImage () {
		if !TrapState.Shared.hasActiveSnapshotImage {
			return
		}
		
		if incrementImage == nil {
			redrawBitmap()
		}
	}
	
	
	
	override func drawRect(rect: CGRect) {
		
		if !TrapState.Shared.hasActiveSnapshotImage {
			return
		}
		
		strokeColor.setStroke()
		
		if incrementImage == nil {
			redrawBitmap()
		}
		
		incrementImage?.drawInRect(displayRect)
		
		switch tool {
		case .Marker, .Arrow, .Circle, .Square:
			
			bzPath.stroke()
			
			if tool == Annotate.Tool.Arrow {
				strokeColor.setFill()
				bzPath.fill()
			}
			
		case .Callout, .Color: break
		}
	}
	

    override func touchesBegan(touches: Set<UITouch>, withEvent event: UIEvent?) {
		
		if !TrapState.Shared.hasActiveSnapshotImage {
			return
		}
		
		touchesDidMove = false
		
		switch tool {
		case .Marker:
			
			UIGraphicsBeginImageContextWithOptions(bufferRect.size, true, 0.0)
			
			incrementImage?.drawInRect(bufferRect)

			strokeColor.setStroke()
			
			counter = 0
			

            markerPoints[0] = touches.first?.locationInView(bufferImageView) ?? CGPointZero
			
		case .Arrow, .Circle, .Square, .Callout:
			
			pathStart = touches.first?.locationInView(bufferImageView) ?? CGPointZero
			
		case .Color: break
		}
	}
	
	
    override func touchesMoved(touches: Set<UITouch>, withEvent event: UIEvent?) {
		
		if !TrapState.Shared.hasActiveSnapshotImage {
			return
		}
		
		touchesDidMove = true
		
		let point = touches.first?.locationInView(bufferImageView) ?? CGPointZero
		
		let difX = point.x - pathStart.x
		let difY = point.y - pathStart.y
		
		var path : UIBezierPath?
		
		switch tool {
		case .Callout, .Color: return
			
		case .Marker:
			
			counter++
			
			markerPoints[counter] = point
			
			if counter == 4 {
				
				// move the endpoint to the middle of the line joining the second control point of the first Bezier segment and the first control point of the second Bezier segment
				markerPoints[3] = CGPoint(x: (markerPoints[2].x + markerPoints[4].x) / 2.0, y: (markerPoints[2].y + markerPoints[4].y) / 2.0)
				
				path = UIBezierPath()
				
				path?.moveToPoint(markerPoints[0])
				
				path?.addCurveToPoint(markerPoints[3], controlPoint1: markerPoints[1], controlPoint2: markerPoints[2])
				
				path?.lineWidth = bzPath.lineWidth
				
				// add to cached path to draw and add to undo stack later
				bzPath.appendPath(path!)
				
				// replace points and get ready to handle the next segment
				markerPoints[0] = markerPoints[3]
				markerPoints[1] = markerPoints[4]
				
				counter = 1
			}
			
			path?.stroke()
			bufferImageView?.image = UIGraphicsGetImageFromCurrentImageContext()
			
		case .Arrow:
			
			let total = hypot(difX, difY)
			
			let tailL = total * 0.7
			
			let tailW = bzPath.lineWidth / 2.0
			
			let headW = (total * 0.2) / 2.0
			
			arrowPoints[0] = CGPoint(x: 0.0,	y: tailW)
			arrowPoints[1] = CGPoint(x: tailL,	y: tailW)
			arrowPoints[2] = CGPoint(x: tailL,	y: headW)
			arrowPoints[3] = CGPoint(x: total,	y: 0.0)
			arrowPoints[4] = CGPoint(x: tailL,	y: -headW)
			arrowPoints[5] = CGPoint(x: tailL,	y: -tailW)
			arrowPoints[6] = CGPoint(x: 0.0,	y: -tailW)
			
			let cosine = difX / total, sine = difY / total
			
			var transform = CGAffineTransform(a: cosine, b: sine, c: -sine, d: cosine, tx: pathStart.x, ty: pathStart.y)
			
			let cgPath = CGPathCreateMutable()
			
			CGPathAddLines(cgPath, &transform, arrowPoints, 7)
			
			CGPathCloseSubpath(cgPath)
			
			path = UIBezierPath(CGPath: cgPath)

			path?.lineWidth = bzPath.lineWidth

			// cache path to draw and add to undo stack later
			bzPath = path!

			
			
			UIGraphicsBeginImageContextWithOptions(bufferRect.size, false, 0.0)
			
			strokeColor.setFill()
			path?.fill()
				
			strokeColor.setStroke()
			path?.stroke()
				
			bufferImageView?.image = UIGraphicsGetImageFromCurrentImageContext()
				
			UIGraphicsEndImageContext()
			
		case .Circle, .Square:
			
			let origin = CGPoint(x: min(pathStart.x, point.x), y: min(pathStart.y, point.y))
			
			let size = CGSize(width: fabs(point.x - pathStart.x), height: fabs(point.y - pathStart.y))
			
			let rect = CGRect(origin: origin, size: size)
			
			path = tool == Annotate.Tool.Circle ? UIBezierPath(ovalInRect: rect) : UIBezierPath(rect: rect)
			
			path?.lineWidth = bzPath.lineWidth

			// cache path to draw and add to undo stack later
			bzPath = path!

			
			
			UIGraphicsBeginImageContextWithOptions(bufferRect.size, false, 0.0)
				
			strokeColor.setStroke()
			path?.stroke()
				
			bufferImageView?.image = UIGraphicsGetImageFromCurrentImageContext()

			UIGraphicsEndImageContext()
		}
	}
	
	
    override func touchesEnded(touches: Set<UITouch>, withEvent event: UIEvent?) {
		
		if !TrapState.Shared.hasActiveSnapshotImage {
			return
		}
		
		switch tool {
		case .Callout:
			
			let rect = CGRect(origin: CGPoint(x: pathStart.x - 16, y: pathStart.y - 14.0), size: CGSize(width: 32.0, height: 28.0))
			// let rect = CGRect(origin: CGPoint(x: pathStart.x - 3.0, y: pathStart.y - 32.0), size: CGSize(width: 28.0, height: 24.0))
			
			let callout = CalloutAnnotation(count: calloutCount, rect: rect, lineWidth: bzPath.lineWidth, color: strokeColor)
			
			annotationDataCache.append(callout)
			
			redrawBitmap()
			
		case .Marker, .Arrow, .Circle, .Square:
			
			if tool == .Marker {
				UIGraphicsEndImageContext()
			}
			
			if touchesDidMove {

				let annotation = PathAnnotation(tool: tool, path: bzPath, color: strokeColor)
				
				annotationDataCache.append(annotation)
				
				redrawBitmap()
			}
			
		case .Color: break
		}
		
		bzPath.removeAllPoints()
		
		pathStart = CGPointZero
		
		counter = 0
		
		touchesDidMove = false
	}
	
    
    override func touchesCancelled(touches: Set<UITouch>?, withEvent event: UIEvent?) {
        if let touch = touches {
            touchesEnded(touch, withEvent: event)
        }
    }
	
	
	
	func redrawBitmap () {
		
		if !TrapState.Shared.hasActiveSnapshotImage {
			return
		}
		
		let setup = isRedrawingAfterOrientationChange || displayRect == CGRectZero
		
		if setup {
			
			if let imageSize = TrapState.Shared.activeSnapshotImage?.size {
				
				let scale = min(bounds.size.width / imageSize.width, bounds.size.height / imageSize.height)
				
				let scaledSize = CGSizeApplyAffineTransform(imageSize, CGAffineTransformMakeScale(scale, scale))
				
				contextRect = CGRect(origin: CGPointZero, size: imageSize)
				
				displayRect = CGRect(origin: CGPoint(x: bounds.midX - scaledSize.width / 2, y: bounds.midY - scaledSize.height / 2), size: scaledSize)
				
				bufferRect = CGRect(origin: CGPointZero, size: scaledSize)
				
				let displayScale = max(imageSize.width / bounds.size.width, imageSize.height / bounds.size.height)
				
				displayTransform = CGAffineTransformMakeScale(displayScale, displayScale)
			}
			
			bufferImageView = UIImageView(frame: displayRect)
			bufferImageView?.backgroundColor = Colors.Clear.uiColor
			addSubview(bufferImageView!)
		}
		
		
		UIGraphicsBeginImageContextWithOptions(contextRect.size, true, 1.0)
		
		strokeColor.setStroke()
		
		if !(incrementImage?.drawAtPoint(CGPointZero) != nil) {
			
			TrapState.Shared.activeSnapshotImage?.drawInRect(contextRect)
		}
		
		
		if annotateUndoManager.redoing {
			
			if let undo = annotationDataCacheArchive.last {
				annotationDataCacheArchive.removeLast()
				annotationDataCache.append(undo)
			}
		}
		
		
		if annotateUndoManager.undoing {
			
			TrapState.Shared.activeSnapshotImage?.drawInRect(contextRect)
			
			if let redo = annotationDataCache.last {
				annotationDataCache.removeLast()
				annotationDataCacheArchive.append(redo)
			}
		}
		
		if !isRedrawingAfterOrientationChange {
		
			let context = UIGraphicsGetCurrentContext()

			CGContextSaveGState(context)

			CGContextConcatCTM(context, displayTransform!)

			if annotateUndoManager.undoing {
				
				calloutCount = 0
				
				for data in annotationDataCache {
					
					switch data.tool {
					case .Callout:
						
						if let callout = data as? CalloutAnnotation {
							
							drawCallout(context, callout: callout)
						}
						
					case .Marker, .Arrow, .Circle, .Square:
						
						if let annotation = data as? PathAnnotation {
							annotation.color.setStroke()
							annotation.path.stroke()
							
							if data.tool == Annotate.Tool.Arrow {
								annotation.color.setFill()
								annotation.path.fill()
							}
						}
						
					case .Color: break
					}
				}
				
			} else if let data = annotationDataCache.last {  // This takes care of redo

				switch data.tool {
				case .Callout:
					
					if let callout = data as? CalloutAnnotation {
						
						drawCallout(context, callout: callout)
					}
					
				case .Marker, .Arrow, .Circle, .Square:
					
					if let annotation = data as? PathAnnotation {
						annotation.color.setStroke()
						annotation.path.stroke()
						
						if data.tool == Annotate.Tool.Arrow {
							annotation.color.setFill()
							annotation.path.fill()
						}
					}
					
				case .Color: break
				}
			}
			
			CGContextRestoreGState(context)
			
			annotateUndoManager.registerUndoWithTarget(self, selector: "redrawBitmap", object: nil)
			
		}
		
		isRedrawingAfterOrientationChange = false
		
		incrementImage = UIGraphicsGetImageFromCurrentImageContext()
		
		UIGraphicsEndImageContext()
		
		bufferImageView?.image = nil
		
		setNeedsDisplay()
	}
	
	
	func drawCallout(context: CGContextRef, callout: CalloutAnnotation) {
		
		callout.color.setStroke()
		
		CGContextSaveGState(context)
		
		callout.clip.addClip()
		callout.path.stroke()
		
		CGContextRestoreGState(context)
		
		callout.tail.stroke()
		
		drawLabel(context, rect: displayRect, yCoord: displayRect.height - callout.center.y, xCoord: callout.center.x, alignment: CTTextAlignment.Center, color: callout.color, label: callout.getCalloutString(), flipContext: true)
		
		calloutCount += 1
	}
	
	
	func drawLabel(ctx: CGContextRef, rect: CGRect, yCoord: CGFloat, xCoord: CGFloat, alignment: CTTextAlignment, color: UIColor, label: String, flipContext: Bool = true) {
		
		let attrString =  NSMutableAttributedString(string: label)
		
		let range = CFRangeMake(0, attrString.length)
		
		if let uiFont = UIFont(name: "HelveticaNeue-Medium", size: range.length > 1 ? 15 : 22) {
		
			let font = CTFontCreateWithName((uiFont.fontName as NSString) as CFString, uiFont.pointSize, nil)
		
			let path = CGPathCreateMutable()

			let alignStyle = NSMutableParagraphStyle()
				alignStyle.alignment = NSTextAlignmentFromCTTextAlignment(alignment)
		
			let attributes : [String : AnyObject] = [ NSFontAttributeName : font, NSForegroundColorAttributeName : color.CGColor, NSParagraphStyleAttributeName : alignStyle ]
				attrString.setAttributes(attributes, range: NSMakeRange(0, attrString.length))
			let target = CGSize(width: CGFloat.max, height: CGFloat.max)

			var fit = CFRange(location: 0, length: 0)
		
			let framesetter = CTFramesetterCreateWithAttributedString(attrString)
		
			let frameSize = CTFramesetterSuggestFrameSizeWithConstraints(framesetter, range, nil, target, &fit)
		
			let textRect = CGRect(x: xCoord - (frameSize.width / 2), y: yCoord - (frameSize.height / 2), width: frameSize.width, height: frameSize.height)
		
			CGPathAddRect(path, nil, textRect)
		
			let frame = CTFramesetterCreateFrame(framesetter, range, path, nil)
		
			if flipContext {
			
				CGContextSaveGState(ctx)
			
				CGContextTranslateCTM(ctx, 0.0, rect.height)
			
				CGContextScaleCTM(ctx, 1, -1)
			
				CTFrameDraw(frame, ctx)
			
				CGContextRestoreGState(ctx)
			} else {
				CTFrameDraw(frame, ctx)
			}
		}
	}
}

File: iOS/Code/Xamarin/bugTrap/bugTrapKit/Controllers/ViewControllers/Views/BtAnnotateImageView.cs


using System;

using Foundation;
using UIKit;
using System.Collections.Generic;
using CoreGraphics;
using CoreText;
using CoreAnimation;
using System.Linq;
using ObjCRuntime;

namespace bugTrapKit
{
	public partial class BtAnnotateImageView : UIView
	{
		public Annotate.Tool Tool = Annotate.Tool.Marker;

		public NSUndoManager AnnotateUndoManager = new NSUndoManager ();

		bool annotationDataCacheHasChangesSinceChecked;

		List<IAnnotationData> annotationDataCache = new List<IAnnotationData> ();
		// observable collection?

		List<IAnnotationData> annotationDataCacheArchive = new List<IAnnotationData> ();

		UIColor strokeColor = Annotate.Color.Black.color();

		UIBezierPath bzPath = new UIBezierPath ();

		CGPoint[] arrowPoints = new CGPoint[7], markerPoints = new CGPoint[5];

		CGPoint pathStart = CGPoint.Empty;

		int counter, calloutCount;

		public UIImage IncrementImage { get; set; }

		// UIImage bufferImage;

		UIImageView bufferImageView;

		bool touchesDidMove;

		CGAffineTransform displayTransform;

		CGRect displayRect = CGRect.Empty, contextRect = CGRect.Empty, bufferRect = CGRect.Empty;

		bool isRedrawingAfterOrientationChange;

		public bool ImageHasChanges {
			get {
				var changed = !annotationDataCache.IsEmpty() && annotationDataCacheHasChangesSinceChecked;
				annotationDataCacheHasChangesSinceChecked = false;
				return changed;
			}
		}


		public BtAnnotateImageView (IntPtr handle) : base(handle)
		{
			MultipleTouchEnabled = false;
			BackgroundColor = UIColor.White;

			ConfigureStroke(Annotate.Color.Black.color());
		}


		public void ResetContextAndState ()
		{
			AnnotateUndoManager.RemoveAllActions(this);
			annotationDataCache = new List<IAnnotationData> ();
			annotationDataCacheArchive = new List<IAnnotationData> ();
			bzPath.RemoveAllPoints();
			arrowPoints = new CGPoint[7];
			markerPoints = new CGPoint[5];
			pathStart = CGPoint.Empty;
			counter = 0;
			IncrementImage = null;
			touchesDidMove = false;
			displayTransform = new CGAffineTransform ();
			displayRect = CGRect.Empty;
			contextRect = CGRect.Empty;
			bufferRect = CGRect.Empty;
		}

		public void UpdateContextAndStateForRotation ()
		{
			displayRect = CGRect.Empty;
			contextRect = CGRect.Empty;
			bufferRect = CGRect.Empty;
			bufferImageView.RemoveFromSuperview();
		}


		public void RefreshLayoutAndImageForRotation ()
		{
			if (!TrapState.Shared.HasActiveSnapshotImage) return;

			isRedrawingAfterOrientationChange = true;

			redrawBitmap();
		}

		public void ConfigureTool (Annotate.Tool annotateTool)
		{
			Tool = annotateTool;
		}

		public void ConfigureStroke (UIColor color, nfloat width = default(nfloat))
		{
			strokeColor = color;

			if (width > 0) bzPath.LineWidth = width;
		}

		public void RefreshLayoutAndImage ()
		{
			if (!TrapState.Shared.HasActiveSnapshotImage) return;

			if (IncrementImage == null) redrawBitmap();
		}

		public override void Draw (CGRect rect)
		{
			//base.Draw(rect);

			if (!TrapState.Shared.HasActiveSnapshotImage) return;

			strokeColor.SetStroke();

			if (IncrementImage == null) redrawBitmap();

			IncrementImage.Draw(displayRect);

			switch (Tool) {
			case Annotate.Tool.Marker:
			case Annotate.Tool.Arrow:
			case Annotate.Tool.Circle:
			case Annotate.Tool.Square:

				bzPath.Stroke();

				if (Tool == Annotate.Tool.Arrow) {
					strokeColor.SetFill();
					bzPath.Fill();
				}

				break;
			case Annotate.Tool.Callout:
			case Annotate.Tool.Color:
				break;
			}
		}

		public override void TouchesBegan (NSSet touches, UIEvent evt)
		{
			// base.TouchesBegan(touches, evt);

			if (!TrapState.Shared.HasActiveSnapshotImage) return;

			touchesDidMove = false;

			var touch = touches.AnyObject as UITouch;

			switch (Tool) {
			case Annotate.Tool.Marker:

				UIGraphics.BeginImageContextWithOptions(bufferRect.Size, true, 0);

				IncrementImage.Draw(bufferRect);

				strokeColor.SetStroke();

				counter = 0;

				markerPoints[0] = touch.LocationInView(bufferImageView);

				break;
			case Annotate.Tool.Arrow:
			case Annotate.Tool.Circle:
			case Annotate.Tool.Square:
			case Annotate.Tool.Callout:

				pathStart = touch.LocationInView(bufferImageView);

				break;
			case Annotate.Tool.Color:
				break;
			}
		}



		public override void TouchesMoved (NSSet touches, UIEvent evt)
		{
			// base.TouchesMoved(touches, evt);

			if (!TrapState.Shared.HasActiveSnapshotImage) return;

			touchesDidMove = true;

			var touch = touches.AnyObject as UITouch;

			var point = touch.LocationInView(bufferImageView);

			var difX = point.X - pathStart.X;
			var difY = point.Y - pathStart.Y;

			var path = new UIBezierPath ();

			switch (Tool) {
			case Annotate.Tool.Callout:
			case Annotate.Tool.Color:

				return;

			case Annotate.Tool.Marker:

				counter++;

				markerPoints[counter] = point;

				if (counter == 4) {

					// move the endpoint to the middle of the line joining the second control point 
					// of the first Bezier segment and the first control point of the second Bezier segment
					markerPoints[3] = new CGPoint ((markerPoints[2].X + markerPoints[4].X) / 2, (markerPoints[2].Y + markerPoints[4].Y) / 2);

					path = new UIBezierPath ();

					path.MoveTo(markerPoints[0]);

					path.AddCurveToPoint(markerPoints[3], markerPoints[1], markerPoints[2]);

					path.LineWidth = bzPath.LineWidth;

					// add to cached path to draw and add to undo stack
					bzPath.AppendPath(path);

					// replace points and get ready to handle the next segment
					markerPoints[0] = markerPoints[3];
					markerPoints[1] = markerPoints[4];

					counter = 1;
				}

				path.Stroke();
				bufferImageView.Image = UIGraphics.GetImageFromCurrentImageContext();

				break;
			case Annotate.Tool.Arrow:

				var total = Math.Sqrt(Math.Pow(difX, 2) + Math.Pow(difX, 2));

				var tailL = (nfloat)(total * 0.7);

				var tailW = bzPath.LineWidth / 2;

				var headW = (nfloat)((total * 0.2) / 2);

				arrowPoints[0] = new CGPoint (0, tailW);
				arrowPoints[1] = new CGPoint (tailL, tailW);
				arrowPoints[2] = new CGPoint (tailL, headW);
				arrowPoints[3] = new CGPoint ((nfloat)total, 0);
				arrowPoints[4] = new CGPoint (tailL, -headW);
				arrowPoints[5] = new CGPoint (tailL, -tailW);
				arrowPoints[6] = new CGPoint (0, -tailW);

				nfloat cosine = (nfloat)(difX / total), sine = (nfloat)(difY / total);

				var transform = new CGAffineTransform (cosine, sine, -sine, cosine, pathStart.X, pathStart.Y);

				var cgPath = new CGPath ();

				cgPath.AddLines(transform, arrowPoints, 7);

				cgPath.CloseSubpath();


				path = UIBezierPath.FromPath(cgPath);

				path.LineWidth = bzPath.LineWidth;

				// cache path to draw and add to undo stack later
				bzPath = path;


				UIGraphics.BeginImageContextWithOptions(bufferRect.Size, false, 0);

				strokeColor.SetFill();
				path.Fill();

				strokeColor.SetStroke();
				path.Stroke();

				bufferImageView.Image = UIGraphics.GetImageFromCurrentImageContext();

				UIGraphics.EndImageContext();

				break;
			case Annotate.Tool.Circle:
			case Annotate.Tool.Square:

				var origin = new CGPoint ((nfloat)Math.Min(pathStart.X, point.X), (nfloat)Math.Min(pathStart.Y, point.Y));

				var size = new CGSize ((nfloat)Math.Abs(difX), (nfloat)Math.Abs(difY));

				var rect = new CGRect (origin, size);

				path = Tool == Annotate.Tool.Circle ? UIBezierPath.FromOval(rect) : UIBezierPath.FromRect(rect);

				path.LineWidth = bzPath.LineWidth;

				// cache path to draw and add to undo stack later
				bzPath = path;


				UIGraphics.BeginImageContextWithOptions(bufferRect.Size, false, 0);

				strokeColor.SetStroke();

				path.Stroke();

				bufferImageView.Image = UIGraphics.GetImageFromCurrentImageContext();

				UIGraphics.EndImageContext();

				break;
			}
		}



		public override void TouchesEnded (NSSet touches, UIEvent evt)
		{
			// base.TouchesEnded(touches, evt);

			if (!TrapState.Shared.HasActiveSnapshotImage) return;


			switch (Tool) {
			case Annotate.Tool.Callout:

				var rect = new CGRect (new CGPoint (pathStart.X - 16, pathStart.Y - 14), new CGSize (32, 28));
				// let rect = CGRect(origin: CGPoint(x: pathStart.x - 3, y: pathStart.y - 32), size: CGSize(width: 28, height: 24))

				var callout = new CalloutAnnotation (calloutCount, rect, bzPath.LineWidth, strokeColor);

				annotationDataCache.Add(callout);

				redrawBitmap();

				break;
			case Annotate.Tool.Marker:
			case Annotate.Tool.Arrow:
			case Annotate.Tool.Circle:
			case Annotate.Tool.Square:

				if (Tool == Annotate.Tool.Marker) UIGraphics.EndImageContext();
				
				if (touchesDidMove) {

					var annotation = new PathAnnotation (Tool, bzPath, strokeColor);

					annotationDataCache.Add(annotation);

					redrawBitmap();
				}

				break;
			case Annotate.Tool.Color:
				break;
			}

			bzPath.RemoveAllPoints();

			pathStart = CGPoint.Empty;

			counter = 0;

			touchesDidMove = false;
		}



		public override void TouchesCancelled (NSSet touches, UIEvent evt)
		{
			TouchesEnded(touches, evt);
			// base.TouchesCancelled(touches, evt);
		}


		[Export("redrawBitmap")]
		void redrawBitmap ()
		{
			if (!TrapState.Shared.HasActiveSnapshotImage) return;

			annotationDataCacheHasChangesSinceChecked = true;

			var setup = isRedrawingAfterOrientationChange || displayRect == CGRect.Empty;

			if (setup) {

				var imageSize = TrapState.Shared.ActiveSnapshotImage.Size;

				var scale = (nfloat)Math.Min(Bounds.Size.Width / imageSize.Width, Bounds.Size.Height / imageSize.Height);

				var scaledSize = CGAffineTransform.MakeScale(scale, scale).TransformSize(imageSize);

				contextRect = new CGRect (CGPoint.Empty, imageSize);

				displayRect = new CGRect (new CGPoint (Bounds.GetMidX() - scaledSize.Width / 2, Bounds.GetMidY() - scaledSize.Height / 2), scaledSize);

				bufferRect = new CGRect (CGPoint.Empty, scaledSize);

				var displayScale = (nfloat)Math.Max(imageSize.Width / Bounds.Size.Width, imageSize.Height / Bounds.Size.Height);

				displayTransform = CGAffineTransform.MakeScale(displayScale, displayScale);


				bufferImageView = new UIImageView (displayRect);
				bufferImageView.BackgroundColor = Colors.Clear;
				AddSubview(bufferImageView);
			}

			UIGraphics.BeginImageContextWithOptions(contextRect.Size, true, 1);

			strokeColor.SetStroke();

			if (IncrementImage != null) {

				IncrementImage.Draw(CGPoint.Empty);
			} else {

				TrapState.Shared.ActiveSnapshotImage.Draw(contextRect);
			}


			if (AnnotateUndoManager.IsRedoing) {

				var undo = annotationDataCacheArchive.LastOrDefault();
				if (undo != null) {
					annotationDataCacheArchive.Remove(undo);
					annotationDataCache.Add(undo);
				}
			}


			if (AnnotateUndoManager.IsUndoing) {

				TrapState.Shared.ActiveSnapshotImage.Draw(contextRect);

				var redo = annotationDataCache.LastOrDefault();
				if (redo != null) {
					annotationDataCache.Remove(redo);
					annotationDataCacheArchive.Add(redo);
				}
			}


			if (!isRedrawingAfterOrientationChange) {

				var context = UIGraphics.GetCurrentContext();

				context.SaveState();

				context.ConcatCTM(displayTransform);

				if (AnnotateUndoManager.IsUndoing) {

					calloutCount = 0;

					foreach (var data in annotationDataCache) {

						switch (data.Tool) {
						case Annotate.Tool.Callout:

							var callout = data as CalloutAnnotation;

							if (callout != null) {
								drawCallout(context, callout);
							}

							break;
						case Annotate.Tool.Marker:
						case Annotate.Tool.Arrow:
						case Annotate.Tool.Circle:
						case Annotate.Tool.Square:

							var annotation = data as PathAnnotation;

							if (annotation != null) {
								annotation.Color.SetStroke();
								annotation.Path.Stroke();

								if (data.Tool == Annotate.Tool.Arrow) {
									annotation.Color.SetFill();
									annotation.Path.Fill();
								}
							}

							break;
						case Annotate.Tool.Color:
							break;
						}
					}

				} else {

					var data = annotationDataCache.LastOrDefault(); // This takes care of redo

					if (data != null) {
					
						switch (data.Tool) {
						case Annotate.Tool.Callout:

							var callout = data as CalloutAnnotation;

							if (callout != null) {
								drawCallout(context, callout);
							}

							break;
						case Annotate.Tool.Marker:
						case Annotate.Tool.Arrow:
						case Annotate.Tool.Circle:
						case Annotate.Tool.Square:

							var annotation = data as PathAnnotation;

							if (annotation != null) {
								annotation.Color.SetStroke();
								annotation.Path.Stroke();

								if (data.Tool == Annotate.Tool.Arrow) {
									annotation.Color.SetFill();
									annotation.Path.Fill();
								}
							}

							break;
						case Annotate.Tool.Color:
							break;
						}
					}
				}

				context.RestoreState();
				
				AnnotateUndoManager.RegisterUndoWithTarget(this, new Selector ("redrawBitmap"), NSString.Empty);
			}

			isRedrawingAfterOrientationChange = false;

			IncrementImage = UIGraphics.GetImageFromCurrentImageContext();

			UIGraphics.EndImageContext();

			bufferImageView.Image = null;

			SetNeedsDisplay();
		}


		void drawCallout (CGContext context, CalloutAnnotation callout)
		{

			callout.Color.SetStroke();

			context.SaveState();

			callout.Clip.AddClip();
			callout.Path.Stroke();

			context.RestoreState();

			callout.Tail.Stroke();

			drawLabel(context, displayRect, displayRect.Height - callout.Center.Y, callout.Center.X, CTTextAlignment.Center, callout.Color, callout.CalloutString, true);

			calloutCount += 1;
		}


		void drawLabel (CGContext ctx, CGRect rect, nfloat yCoord, nfloat xCoord, CTTextAlignment alignment, UIColor color, string label, bool flipContext = true)
		{

			Console.WriteLine("Finish Draw code");

			var attrString = new NSMutableAttributedString (label);

			var range = new NSRange (0, attrString.Length);

			var uiFont = UIFont.FromName("HelveticaNeue-Medium", range.Length > 1 ? 15 : 22);

			var font = new CTFont (uiFont.Name, uiFont.PointSize);  //((uiFont.Name as NSString) as CFString, uiFont.pointSize, nil);

			var path = new CGPath ();

			var alignStyle = new CTParagraphStyle (new CTParagraphStyleSettings { Alignment = alignment });

			var attributes = new CTStringAttributes {
				Font = font,
				ForegroundColor = color.CGColor,
				ParagraphStyle = alignStyle
			};

			attrString.SetAttributes(attributes, new NSRange (0, attrString.Length));

			var target = new CGSize (nfloat.MaxValue, nfloat.MaxValue);

			var fit = new NSRange (0, 0);

			var framesetter = new CTFramesetter (attrString);

			var frameSize = framesetter.SuggestFrameSize(range, null, target, out fit);

			var textRect = new CGRect (xCoord - (frameSize.Width / 2), yCoord - (frameSize.Height / 2), frameSize.Width, frameSize.Height);

			path.AddRect(textRect);

			var frame = framesetter.GetFrame(range, path, null);

			if (flipContext) {

				ctx.SaveState();

				ctx.TranslateCTM(0, rect.Height);

				ctx.ScaleCTM(1, -1);

				frame.Draw(ctx);

				ctx.RestoreState();

			} else {
				frame.Draw(ctx);
			}
		}
	}
}
